<a href="https://colab.research.google.com/github/navronaman/RAISE2024-Bloustein/blob/main/INDIA_RAISE2024_AI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# RAISE 2024 - Team India
Team Members: Avani, Naman, Prisha, Tanay
# Loading the dataset

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import pandas as pd
import numpy as np
from tqdm.notebook import tqdm
path = "/content/drive/MyDrive/Spring 2024/Hackathons/Dataset_10k.csv"
df = pd.read_csv(path)
df.head()

title  \
0  Kann künstliche Intelligenz Smalltalk ersetzen...   
1  Lidé se snadněji svěří chatbotům. Emocionální ...   
2  ¿Cómo implementar la IA en el comercio online?...   
3  Kontrolle von KI: Vorsicht, KI! | ZEIT ONLINE ...   
4  Los consejos para ganar la lotería de la Intel...   

                                                link        date  \
0  https://news.google.com/rss/articles/CBMiY2h0d...  2023-11-12   
1  https://news.google.com/rss/articles/CBMidmh0d...  2023-11-12   
2  https://news.google.com/rss/articles/CBMicmh0d...  2023-11-12   
3  https://news.google.com/rss/articles/CBMiXWh0d...  2023-11-12   
4  https://news.google.com/rss/articles/CBMiXWh0d...  2023-11-12   

                source country language  \
0                Blick      US       de   
1             iDNES.cz      US       cs   
2         RRHH Digital      US       es   
3  ZEIT ONLINE English      US       de   
4                  C5N      US       es   

                                    translated_title  \
0  Can artificial intelligence replace small talk...   
1  People trust chatbots more easily. Emotional A...   
2  How to implement AI in online commerce? - Digi...   
3  Controlling AI: Be careful, AI! | ZEIT ONLINE ...   
4  Tips to win the Artificial Intelligence lotter...   

   number_of_characters_title  number_of_words_title day_of_week  ...  \
0                          47                      8      Sunday  ...   
1                          72                     13      Sunday  ...   
2                          43                     10      Sunday  ...   
3                          56                     12      Sunday  ...   
4                          45                      9      Sunday  ...   

   is_weekend  is_holiday  source_type  \
0        True       False  News Outlet   
1        True       False        Other   
2        True       False  News Outlet   
3        True       False        Other   
4        True       False        Other   

                                final_redirected_URL        domain_of_URL  \
0  https://www.blick.ch/schweiz/linguistik-profes...         www.blick.ch   
1  https://www.idnes.cz/zpravy/domaci/ai-umela-in...         www.idnes.cz   
2  https://www.rrhhdigital.com/secciones/tecnolog...  www.rrhhdigital.com   
3  https://www.zeit.de/zeit-wissen/2023/06/kuenst...          www.zeit.de   
4  https://www.c5n.com/sociedad/los-consejos-gana...          www.c5n.com   

  subdomain_of_URL URL_depth top_level_domain url_length     author  
0              www         2               ch         99         []  
1              www         3               cz        118  ['Autor']  
2              www         5              com        114         []  
3              www         4               de         93         []  
4              www         2              com         93         []  

[5 rows x 23 columns]

In [4]:
import matplotlib.pyplot as plt
import seaborn as sns

plt.style.use('ggplot')

import nltk

# Performing EDA


Getting the column names


In [5]:
df.columns

Index(['title', 'link', 'date', 'source', 'country', 'language',
       'translated_title', 'number_of_characters_title',
       'number_of_words_title', 'day_of_week', 'month', 'year', 'quarter',
       'is_weekend', 'is_holiday', 'source_type', 'final_redirected_URL',
       'domain_of_URL', 'subdomain_of_URL', 'URL_depth', 'top_level_domain',
       'url_length', 'author'],
      dtype='object')

In [6]:
print(df['country'].unique())
print(df['language'].unique())
print(df['year'].unique())
print(df['source_type'].unique())
print(df['top_level_domain'].unique())
print(df['domain_of_URL'].unique())
print(len(df['domain_of_URL'].unique()))

['US']
['de' 'cs' 'es' 'hu' 'id' 'vi' 'fr' 'ro' 'pl' 'en' 'pt' 'ko' 'hi' 'ja'
 'zh-cn' 'tr' 'it' 'zh_Hant' 'bn' 'no' 'ru' 'th' 'el' 'et' 'ca' 'sv' 'fa'
 'sl' 'ar' 'af' 'sr' 'zh' 'iw' 'nl' 'uk' 'da' 'pa' 'sw' 'hr' 'fi' 'sk'
 'he' 'te' 'mk' 'ml' 'mr' 'gu' 'bg' 'so' 'zh-tw' 'tl' 'lv']
[2023]
['News Outlet' 'Other' 'Academic Journal' 'Press Release Service' 'Blog'
 'Broadcast Media' 'Magazine' 'Online Forum' 'Corporate']
['ch' 'cz' 'com' 'de' 'hu' nan 'es' 'info' 'uk' 'br' 'kr' 'pt' 'mx' 'vn'
 'org' 'pe' 'in' 'capetown' 'jp' 'net' 'tr' 'cl' 'it' 'tw' 'eu' 'sg' 'io'
 'nl' 'th' 'at' 'au' 'news' 'cn' 'asia' 'st' 'my' 'ie' 'gr' 'id' 'bn' 'ar'
 'uy' 'ru' 'cat' 'se' 'no' 'si' 'ca' 'fr' 'tv' 'co' 'do' 'ai' 'aero' 'so'
 'nz' 'za' 'edu' 'int' 'rs' 'il' 'media' 'network' 'zw' 'hk' 'biz' 'world'
 'ua' 'pl' 'healthcare' 'gov' 'mil' 'na' 'ug' 'ng' 'press' 'hr' 'fi' 'sk'
 'ly' 'ro' 'africa' 'fj' 'bg' 'today' 'fyi' 'ke' 'sa' 'global' 'hamburg'
 'pk' 'tn' 'lu' 'us' 'im' 'ventures' 'be' 'works' 'review' 'g

In [7]:
print(len(df))
print(len(df[df['language'] == 'en']))
print(len(df[df['source_type'] == 'News Outlet']))
print(len(df[df['source_type'] == 'Other']))
print(len(df[df['source_type'] == 'Academic Journal']))
print(len(df[df['source_type'] == 'Press Release Service']))
print(len(df[df['source_type'] == 'Blog']))
print(len(df[df['source_type'] == 'Broadcast Media']))
print(len(df[df['source_type'] == 'Magazine']))
print(len(df[df['source_type'] == 'Online Forum']))
print(len(df[df['source_type'] == 'Corporate']))
temp_df = df[(df['language']=='en') & ((df['source_type']=='News Outlet') | (df['source_type']=='Other') | (df['source_type']=='Blog'))]
print(len(temp_df))

10000
8390
3722
4028
382
394
879
140
408
13
34
7125


We will subset our data to include 7125 articles, all in the English Language, subsetted by source types of News Outlets, Blogs or Other.

In [8]:
df = df[(df['language']=='en') & ((df['source_type']=='News Outlet') | (df['source_type']=='Other') | (df['source_type']=='Blog'))]
df.head()

title  \
19  Time travel and genius AI: Research roundup - ...   
25    Nick Bostrom: Will AI lead to tyranny? - UnHerd   
26  Top 10 AI Coins To Invest In 2024 - NFTevening...   
27  Samsung AI Forum 2023 day 2: Discussing techno...   
28  Read Full SAG-AFTRA Deal Summary Released Ahea...   

                                                 link        date  \
19  https://news.google.com/rss/articles/CBMiJ2h0d...  2023-11-12   
25  https://news.google.com/rss/articles/CBMiQGh0d...  2023-11-12   
26  https://news.google.com/rss/articles/CBMiQGh0d...  2023-11-12   
27  https://news.google.com/rss/articles/CBMiogFod...  2023-11-12   
28  https://news.google.com/rss/articles/CBMiUGh0d...  2023-11-12   

            source country language  \
19  Varsity Online      US       en   
25          UnHerd      US       en   
26  NFTevening.com      US       en   
27           ZAWYA      US       en   
28        Deadline      US       en   

                                     translated_title  \
19  Time travel and genius AI: Research roundup - ...   
25    Nick Bostrom: Will AI lead to tyranny? - UnHerd   
26  Top 10 AI Coins To Invest In 2024 - NFTevening...   
27  Samsung AI Forum 2023 day 2: Discussing techno...   
28  Read Full SAG-AFTRA Deal Summary Released Ahea...   

    number_of_characters_title  number_of_words_title day_of_week  ...  \
19                          51                     10      Sunday  ...   
25                          39                      9      Sunday  ...   
26                          41                     10      Sunday  ...   
27                          84                     17      Sunday  ...   
28                          95                     18      Sunday  ...   

    is_weekend  is_holiday  source_type  \
19        True       False        Other   
25        True       False        Other   
26        True       False        Other   
27        True       False        Other   
28        True       False        Other   

                                 final_redirected_URL      domain_of_URL  \
19            https://www.varsity.co.uk/science/26422  www.varsity.co.uk   
25  https://unherd.com/2023/11/nick-bostrom-will-a...         unherd.com   
26  https://nftevening.com/10-ai-crypto-coins-to-b...     nftevening.com   
27  https://www.zawya.com/en/press-release/events-...      www.zawya.com   
28  https://deadline.com/2023/11/sag-aftra-deal-su...       deadline.com   

   subdomain_of_URL URL_depth top_level_domain url_length  \
19              www         2               uk         39   
25           unherd         4              com         64   
26       nftevening         2              com         64   
27              www         4              com        162   
28         deadline         4              com         80   

                                               author  
19                 ['Daniel Hilton', 'Grace Wakeman']  
25  ['Flo Read', 'Nell Clover', 'Douglas Redmayne'...  
26                                                 []  
27                                  ['Press Release']  
28                                 ['Dominic Patten']  

[5 rows x 23 columns]

# Sentiment Analysis on the Headlines using Roberta Pretrained Model


In [9]:
from transformers import AutoTokenizer
from transformers import AutoModelForSequenceClassification
from scipy.special import softmax

In [10]:
MODEL = f"cardiffnlp/twitter-roberta-base-sentiment"
tokenizer = AutoTokenizer.from_pretrained(MODEL)
model = AutoModelForSequenceClassification.from_pretrained(MODEL)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/747 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/150 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/499M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()


In [23]:
example = df['title'][70]
print(example)

encoded_text = tokenizer(example, return_tensors='pt')
print(encoded_text)

output = model(**encoded_text)
scores = softmax(output[0][0].detach().numpy())
scores_dict = {
    'neg': scores[0],
    'neu': scores[0],
    'pos': scores[0]
}
print(scores_dict)

Bill Gates: AI Is About to Completely Change How You Use Computers - PCMag AU
{'input_ids': tensor([[    0, 19993, 10090,    35,  4687,  1534,  1936,     7, 18337,   352,
          7229,  1336,   370,  7627, 34710,   268,   111,  4985, 25117, 16778,
             2]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}
{'neg': 0.0067375773, 'neu': 0.0067375773, 'pos': 0.0067375773}


In [25]:
def polarity_scores(example):
  encoded_text = tokenizer(example, return_tensors='pt')
  output = model(**encoded_text)
  scores = softmax(output[0][0].detach().numpy())
  scores_dict = {
      'neg': scores[0],
      'neu': scores[0],
      'pos': scores[0]
  }
  return scores_dict

In [29]:
print(df['title'][1235])
print(polarity_scores(df['title'][1235]))

How Generative AI will Change Traditional SaaS Models with AWS - SaaStr
{'neg': 0.006017257, 'neu': 0.006017257, 'pos': 0.006017257}


We notice that with headlines, even the expert roBERTa model is unable to make detect the positivity and negativity.